infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -0,0 +1,133 @@
+package org.corfudb.infrastructure;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.CacheWriter;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.RemovalCause;
+import com.github.benmanes.caffeine.cache.RemovalListener;
+import com.google.common.annotations.VisibleForTesting;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.runtime.view.Address;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Sequencer server cache.
+ * Contains transaction conflict-resolution data structures.
+ */
+@Slf4j
+public class SequencerServerCache {
+    /**
+     * TX conflict-resolution information:
+     * <p>
+     * a cache of recent conflict keys and their latest global-log position.
+     */
+    private final Cache<String, Long> conflictToGlobalTailCache;
+
+    /**
+     * A "wildcard" representing the maximal update timestamp of
+     * all the conflict keys which were evicted from the cache
+     */
+    @Getter
+    private long maxConflictWildcard = Address.NOT_FOUND;
+
+    /**
+     * maxConflictNewSequencer represents the max update timestamp of all the conflict keys
+     * which were evicted from the cache by the time this server is elected
+     * the primary sequencer. This means that any snapshot timestamp below this
+     * actual threshold would abort due to NEW_SEQUENCER cause.
+     */
+    @Getter
+    private long maxConflictNewSequencer = Address.NOT_FOUND;
+
+    @VisibleForTesting
+    public SequencerServerCache(long cacheSize, CacheWriter<String, Long> writer) {
+        this.conflictToGlobalTailCache = Caffeine.newBuilder()
+                .maximumSize(cacheSize)
+                .writer(writer)
+                .executor(Runnable::run)
+                .recordStats()
+                .build();
+    }
+
+    /**
+     * Limited cache by size.
+     * Eviction policy described here https://github.com/ben-manes/caffeine/wiki/Writer
+     * @param cacheSize cache size
+     */
+    public SequencerServerCache(long cacheSize) {
+        CacheWriter<String, Long> writer = new CacheWriter<String, Long>() {
+            @Override
+            public void write(@Nonnull String key, @Nonnull Long value) {
+                //ignore
+            }
+
+            @Override
+            public void delete(@Nonnull String key, @Nullable Long globalAddress, @Nonnull RemovalCause cause) {
+                if (cause == RemovalCause.REPLACED) {
```

Comment:
throw an exception ?

Reply:
Not sure. It was in the prev version, I wouldn't like to change the behaviour in this case

Reply:
After some discussion, we decided to throw IllegalStateException.
Done